### PR TITLE
[py] change "By" class with enum

### PR DESCRIPTION
**Pull request by plming**
### Description
* Refactoring "By" class with standarm Enum library
* Add "By" type hint where it needed.

### Motivation and Context
I checked history for "by.py" file. The file was first committed in 2010.
But according to [PEP 435](https://peps.python.org/pep-0435/), enum is introduced to Python after 2013.
So I think there's no reason to avoid standard enum library.
I expect this PR can enforce programmers not to use magic strings.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed. ;; Fortunately, there's no any piece of codes depending on magic strings.

### CODE REVIEW

File: py/selenium/webdriver/common/by.py [hunk 1]
```diff
@@ -19,17 +19,20 @@
 The By implementation.
 """
 
+from enum import Enum, unique, auto
 
-class By:
+
+@unique
+class By(Enum):
```
Comment (symonk) [inline py/selenium/webdriver/common/by.py]: This is not such a trivial change here

File: py/selenium/webdriver/common/by.py [hunk 2]
```diff
@@ -19,17 +19,20 @@
...
 
-    ID = "id"
-    XPATH = "xpath"
-    LINK_TEXT = "link text"
-    PARTIAL_LINK_TEXT = "partial link text"
-    NAME = "name"
-    TAG_NAME = "tag name"
-    CLASS_NAME = "class name"
-    CSS_SELECTOR = "css selector"
+    ID = auto()
```
Comment (symonk) [inline py/selenium/webdriver/common/by.py]: won't this throw `using:` way off?

Comment (plming) [reply py/selenium/webdriver/common/by.py]: > won't this throw `using:` way off?

I'm afraid but could you explain in detail? Do you mean that constants are used in somewhere?

Comment (symonk) [reply py/selenium/webdriver/common/by.py]: yes, the old `str` value is passed into the `using: by`

**Comment by plming** [comment]
I found there's a code using "json.dumps function" and that function needs constants for dumping..
So I tried to fix enum with str. How do you think about it?

**Comment by Solonotix** [comment]
The failing test appears to be because when the exception for NoSuchElement is created, the effective call is to `repr()` on the resulting dictionary, thus the key looks like `{<By.ID: 'id'>: 'nonexistentid'}`. If the By class had a `__repr__` implementation that returned `self.value` instead, then it seems like it would pass the currently failing test on `test_no_such_element_is_raised_rather_than_index_error`

**Pull request closed without merging**